Feature/anthropic sdk and tools

TheApeMachine opened the pull request:
# Pull Request Template

## Description

This PR enhances the Anthropic integration to make it more aligned with the OpenAI implementation. The changes include adding proper streaming support, implementing multimodal capabilities, improving error handling, and ensuring consistent behavior between the two LLM implementations.

Specifically, this PR addresses:

1. Implemented enhanced `stream` method to correctly handle streaming responses and properly accumulate text content
2. Added support for stop sequences and the `tool_choice` parameter in the `buildChatCompletionRequest` method
3. Improved error handling by wrapping errors with `ErrAnthropicChat` for consistent error reporting
4. Removed debug print statements for cleaner production code
5. Added multimodal support by properly handling image content in the `threadToChatCompletionMessages` function

Fixes #225

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

The changes have been tested using the example code provided in the repository:

- [x] Streaming example (`examples/llm/antropic/stream/main.go`) - Verified that streaming responses are correctly displayed
- [x] Multimodal example (`examples/llm/antropic/multimodal/main.go`) - Verified that image content is correctly processed and described

## Checklist:

- [x] I have read the [Contributing documentation](https://github.com/henomis/lingoose/blob/main/CONTRIBUTING.md).
- [x] I have read the [Code of conduct documentation](https://github.com/henomis/lingoose/blob/main/CODE_OF_CONDUCT.md).
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

TheApeMachine committed:
Update dependencies and refactor Anthropic integration

- Bump Go version to 1.22.0 and update toolchain.
- Add new dependency: `github.com/anthropics/anthropic-sdk-go` v0.2.0-alpha.8.
- Update existing dependencies: `golang.org/x/net` to v0.27.0 and `github.com/stretchr/testify` to v1.9.0.
- Refactor Anthropic client initialization to use API key from environment variable.
- Change method calls from `Generate` to `Chat` for improved functionality.
- Remove deprecated `api.go` and `formatter.go` files; introduce new `common.go` and `function.go` for better structure.
- Enhance message handling and tool integration in the Anthropic client.

Commit:
Enhance Anthropic LLM integration with new features and improvements

- Add support for response format (JSON/text)
- Implement caching mechanism for responses
- Add usage metadata tracking
- Introduce new methods for configuration (WithResponseFormat, SetStop)
- Improve error handling and logging
- Refactor Generate method with more robust generation process

Commit:
Refactor Anthropic LLM streaming and message formatting

- Improve streaming response handling in `stream` method
- Enhance message content processing in `threadToChatCompletionMessages`
- Add support for multi-content messages (text and images)
- Optimize tool choice and tools parameter handling
- Remove redundant token tracking and simplify streaming logic

Commit:
Remove unused delta type constants in Anthropic client

Clean up unused constants related to stream delta types, simplifying the Anthropic client configuration

Commit:
Merge branch 'main' into feature/anthropic-sdk-and-tools

Commit:
merge with main

henomis commented on COMMIT.md:
```diff
@@ -0,0 +1,41 @@
+# Pull Request Template
```
please, remove this file

henomis commented on llm/anthropic/anthropic.go:
```diff
@@ -131,161 +138,364 @@ func (o *Antropic) setCache(ctx context.Context, t *thread.Thread, cacheResult *
-	} else {
-		err = o.generate(ctx, t, chatRequest)
+// NewAnthropic creates a new Anthropic instance with explicit API key.
+func NewAnthropic(apiKey string) *Anthropic {
```
Please implement a WithKey method here

TheApeMachine replied:
I am not 100% sure about my implementation here.

henomis commented on llm/anthropic/anthropic.go:
```diff
@@ -131,161 +138,364 @@ func (o *Antropic) setCache(ctx context.Context, t *thread.Thread, cacheResult *
+}
+
+// Chat implements the LLM interface
+func (a *Anthropic) Chat(ctx context.Context, t *thread.Thread) error {
```
remove this method we should support only `Generate`

henomis commented on examples/embeddings/voyage/main.go:
```diff
@@ -58,7 +59,7 @@ func main() {
 	}
 
-	anthropicllm := anthropic.New().WithModel("claude-3-opus-20240229")
+	anthropicllm := anthropic.NewAnthropic(os.Getenv("ANTHROPIC_API_KEY")).WithModel(anthropic.ModelClaude_3_Opus_20240229)
```
Please do not change the examples. Use `New` here and `Generate`.

henomis commented on llm/anthropic/anthropic.go:
```diff
@@ -131,161 +138,364 @@ func (o *Antropic) setCache(ctx context.Context, t *thread.Thread, cacheResult *
-	if err != nil {
-		return fmt.Errorf("%w: %w", ErrAnthropicChat, err)
+// New creates a new Anthropic instance with API key from environment
+func New() *Anthropic {
```
put `New` on top just under the structure, then all the `With*` methods, then public method, then private methods

henomis commented on examples/llm/antropic/multimodal/main.go:
```diff
@@ -3,13 +3,14 @@ package main
 
 func main() {
-	anthropicllm := anthropic.New().WithModel("claude-3-opus-20240229")
+	anthropicllm := anthropic.NewAnthropic(os.Getenv("ANTHROPIC_API_KEY")).WithModel(anthropic.ModelClaude_3_Opus_20240229)
```
Please provide some examples showing tools/functions usage like the ones here:https://github.com/henomis/lingoose/tree/main/examples/llm/openai/tools
Same for the response format like the one here https://github.com/henomis/lingoose/tree/main/examples/llm/openai/response_format

henomis reviewed (changes requested):
Left some comments, just a quick review, need to go deeper, Great job!

TheApeMachine commented:
I will work on this coming week.

Commit:
Refactor Anthropic client initialization and method calls

- Remove deprecated NewAnthropic method and replace it with a simplified New method that retrieves the API key from the environment.
- Update example usages in the `stream`, `multimodal`, and `voyage` examples to use the new client initialization and method calls.
- Change method calls from `Chat` to `Generate` for improved functionality and consistency across examples.
- Clean up unused imports and improve overall code clarity.

Ran existing examples to confirm they are still working. Will implement new examples as requested this week.

henomis commented on llm/anthropic/anthropic.go:
```diff
@@ -19,90 +17,137 @@ import (
-	EOS               = "\x00"
-)
+// WithKey returns a new Anthropic instance with the given API key.
+func WithKey(key string) *Anthropic {
```
```suggestion
func (a *Anthropic) WithKey(key string) *Anthropic {
```
